src/semidiscretization/semidiscretization.jl

```diff
@@ -72,8 +86,19 @@ function semidiscretize(semi::AbstractSemidiscretization, tspan)
   # TODO: MPI, do we want to synchonize loading and print debug statements, e.g. using
   #       mpi_isparallel() && MPI.Barrier(mpi_comm())
   #       See https://github.com/trixi-framework/Trixi.jl/issues/328
+  wrapper = RHSWrapper(semi, rhs!) # wrap `rhs!`` and `semi` to reduce latency, see `RHSWrapper`
   iip = true # is-inplace, i.e., we modify a vector when calling rhs!
-  return ODEProblem{iip}(rhs!, u0_ode, tspan, semi)
+  return ODEProblem{iip}(wrapper, u0_ode, tspan)
+end
+
+
+# get the semidiscretization from an `ODEIntegrator`
+function extract_semidiscretization(integrator)
+  if integrator.f isa RHSWrapper
+    return integrator.f.semi
+  else
+    return integrator.p
+  end
```

Comment:
OOC (out-of-curiosity): Is this type stable? Or would it make sense to dispatch on the type of `integrator.f`, which is probably a type parameter of `integrator`? Or am I completely off track since we always return a `semi` and thus the return type _is_ stable?

Reply:
It should be type table since it should specialize on the type of `integrator`, which knows the type of `integrator.f` - but it may be worth checking to be sure - and running some benchmarks,of course

Reply:
OK, thanks!

Reply:
The wrapping with losing type information is done at the level of OrdinaryDiffEq.jl. This means that accessing the semidiscretization like this is indeed not type stable. Thus, I can think of the following options

- keep the code as it is in this PR, introducing type instabilities in callbacks accessing `semi`
- introduce a function barrier in all callbacks after accessing `semi` to mitigate this problem
- store the semidiscretization `semi` in the callbacks requiring it (if feasible, i.e., no full duplication, just a pointer)
- keep the status quo without this PR

Reply:
Have you been able to assess the impact on latency/compilation times with this change yet? Maybe we should try to test this on Roci for a more stable test environment?

If it _does_ improve things (in a relevant way) in terms of TTFX, I think keeping the status quo is not desirable. In that case, I would try to go for the least complex, then the least intrusive solution.

Reply:
Please see the post below: https://github.com/trixi-framework/Trixi.jl/pull/1255#issuecomment-1308452741